diff --git a/bin/install-p4dev-v8.sh b/bin/install-p4dev-v8.sh
index eab1410..c1a4e22 100755
--- a/bin/install-p4dev-v8.sh
+++ b/bin/install-p4dev-v8.sh
@@ -1020,28 +1020,33 @@ TIME_MININET_START=$(date +%s)
 # continue to apply.  Will likely want to update this to newer
 # versions once or twice a year.
 MININET_COMMIT="6eb8973c0bfd13c25c244a3871130c5e36b5fbd7"  # 2024-Sep-18
-git clone https://github.com/mininet/mininet mininet
-cd mininet
-git checkout ${MININET_COMMIT}
-PATCH_DIR="${THIS_SCRIPT_DIR_ABSOLUTE}/patches"
-patch -p1 < "${PATCH_DIR}/mininet-patch-for-2024-sep-enable-venv.patch"
-cd ..
-RESTORE_SUDOERS_FILE=0
-if [ -e /etc/sudoers.d/sudoers-dotfiles ]
+if [ -d mininet ]
 then
-    # Starting with Ubuntu 24.04, by default it includes a sudo
-    # configuration file that disallows passing environment variables
-    # such as DEBIAN_FRONTEND on a sudo command line for apt-get
-    # commands.  On such systems, temporarily rename this
-    # configuration file while installing Mininet, since Mininet's
-    # install script uses this feature of sudo.
-    sudo mv /etc/sudoers.d/sudoers-dotfiles /etc/sudoers.d/sudoers-dotfiles.orig
-    RESTORE_SUDOERS_FILE=1
-fi
-PYTHON=python3 ./mininet/util/install.sh -nw
-if [ ${RESTORE_SUDOERS_FILE} -eq 1 ]
-then
-    sudo mv /etc/sudoers.d/sudoers-dotfiles.orig /etc/sudoers.d/sudoers-dotfiles
+    echo "Found directory ${INSTALL_DIR}/mininet.  Assuming desired version of mininet is already installed."
+else
+    git clone https://github.com/mininet/mininet mininet
+    cd mininet
+    git checkout ${MININET_COMMIT}
+    PATCH_DIR="${THIS_SCRIPT_DIR_ABSOLUTE}/patches"
+    patch -p1 < "${PATCH_DIR}/mininet-patch-for-2024-sep-enable-venv.patch"
+    cd ..
+    RESTORE_SUDOERS_FILE=0
+    if [ -e /etc/sudoers.d/sudoers-dotfiles ]
+    then
+        # Starting with Ubuntu 24.04, by default it includes a sudo
+        # configuration file that disallows passing environment variables
+        # such as DEBIAN_FRONTEND on a sudo command line for apt-get
+        # commands.  On such systems, temporarily rename this
+        # configuration file while installing Mininet, since Mininet's
+        # install script uses this feature of sudo.
+        sudo mv /etc/sudoers.d/sudoers-dotfiles /etc/sudoers.d/sudoers-dotfiles.orig
+        RESTORE_SUDOERS_FILE=1
+    fi
+    PYTHON=python3 ./mininet/util/install.sh -nw
+    if [ ${RESTORE_SUDOERS_FILE} -eq 1 ]
+    then
+        sudo mv /etc/sudoers.d/sudoers-dotfiles.orig /etc/sudoers.d/sudoers-dotfiles
+    fi
 fi
 TIME_MININET_END=$(date +%s)
 echo "mininet                : $(($TIME_MININET_END-$TIME_MININET_START)) sec"
@@ -1070,13 +1075,18 @@ TIME_PTF_START=$(date +%s)
 # Ubuntu 22.04.
 #sudo pip3 install pypcap
 
-git clone https://github.com/p4lang/ptf
-cd ptf
-if [ "x${INSTALL_PTF_SOURCE_VERSION}" != "x" ]; then
-    git checkout ${INSTALL_PTF_SOURCE_VERSION}
+if [ -d ptf ]
+then
+    echo "Found directory ${INSTALL_DIR}/ptf.  Assuming desired version of ptf is already installed."
+else
+    git clone https://github.com/p4lang/ptf
+    cd ptf
+    if [ "x${INSTALL_PTF_SOURCE_VERSION}" != "x" ]; then
+        git checkout ${INSTALL_PTF_SOURCE_VERSION}
+    fi
+    git log -n 1
+    pip install .
 fi
-git log -n 1
-pip install .
 TIME_PTF_END=$(date +%s)
 echo "p4lang/ptf             : $(($TIME_PTF_END-$TIME_PTF_START)) sec"
 
@@ -1126,15 +1136,20 @@ else
     pip3 install grpcio==1.51.3
 fi
 
-git clone https://github.com/p4lang/p4runtime-shell
-cd p4runtime-shell
-if [ "x${INSTALL_P4RUNTIME_SHELL_SOURCE_VERSION}" != "x" ]; then
-    git checkout ${INSTALL_P4RUNTIME_SHELL_SOURCE_VERSION}
+if [ -d p4runtime-shell ]
+then
+    echo "Found directory ${INSTALL_DIR}/p4runtime-shell.  Assuming desired version of p4runtime-shell is already installed."
+else
+    git clone https://github.com/p4lang/p4runtime-shell
+    cd p4runtime-shell
+    if [ "x${INSTALL_P4RUNTIME_SHELL_SOURCE_VERSION}" != "x" ]; then
+        git checkout ${INSTALL_P4RUNTIME_SHELL_SOURCE_VERSION}
+    fi
+    git log -n 1
+    PATCH_DIR="${THIS_SCRIPT_DIR_ABSOLUTE}/patches"
+    patch -p1 < "${PATCH_DIR}/p4runtime-shell-2023-changes.patch"
+    pip3 install .
 fi
-git log -n 1
-PATCH_DIR="${THIS_SCRIPT_DIR_ABSOLUTE}/patches"
-patch -p1 < "${PATCH_DIR}/p4runtime-shell-2023-changes.patch"
-pip3 install .
 
 pip3 list
 
@@ -1151,9 +1166,14 @@ echo "start install tutorials:"
 set -x
 date
 
-git clone https://github.com/p4lang/tutorials
-cd tutorials
-git log -n 1
+if [ -d tutorials ]
+then
+    echo "Found directory ${INSTALL_DIR}/tutorials.  Assuming desired version of tutorials is already installed."
+else
+    git clone https://github.com/p4lang/tutorials
+    cd tutorials
+    git log -n 1
+fi
 
 set +x
 echo "end install tutorials:"
